projects/hang_analysis.kp/knowledge.md

```diff
@@ -0,0 +1,334 @@
+    zipped = zip(data.values, map(int, data.keys()))
+    vals = sorted(zipped, key=lambda x: x[1])
+
+    return pd.Series([v for v,k in vals], index=[k for v,k in vals]).truncate(before=2048).sum()
+    
+def count_content_input_delays(ping):
+    if ping['payload/processes/content/histograms/INPUT_EVENT_RESPONSE_COALESCED_MS/values'] is None:
+        return 0
+    data = ping['payload/processes/content/histograms/INPUT_EVENT_RESPONSE_COALESCED_MS/values']
+    zipped = zip(data.values(), map(int, data.keys()))
+    vals = sorted(zipped, key=lambda x: x[1])
+
+    return pd.Series([v for v,k in vals], index=[k for v,k in vals]).truncate(before=2048).sum()
+
+def get_counts(ping):
+    hangs = flatten_all_hangs(ping)
+    subsession_length = ping['payload/info/subsessionLength']
+    return (ping['application/buildId'], {
+        'subsession_length': subsession_length,
+        'parent_bhr': count_bhr_hangs('Gecko', hangs),
+        'content_bhr': count_bhr_hangs('Gecko_Child', hangs),
+        'parent_input': count_parent_input_delays(ping),
+        'content_input': count_content_input_delays(ping),
+    })
+
+def merge_counts(a, b):
+    return {k: a[k] + b[k] for k in a.iterkeys()}
+
+def ping_is_valid(ping):
+    if not isinstance(ping["application/buildId"], basestring):
+        return False
+    if type(ping["payload/info/subsessionLength"]) != int:
+        return False
+
+    return ping["environment/system/os/name"] == "Windows_NT"
+
+cached = subset.filter(ping_is_valid).map(get_counts).cache()
+counts_result = cached.reduceByKey(merge_counts).collect()
+```
+
+```python
+sorted_result = sorted(counts_result, key=lambda x: x[1])
+```
+
+```python
+plot_data = np.array([
+    [float(x['parent_bhr']) / x['subsession_length'] * 3600. * 1000. for k,x in sorted_result],
+    [float(x['content_bhr']) / x['subsession_length'] * 3600. * 1000. for k,x in sorted_result],
+    [float(x['parent_input']) / x['subsession_length'] * 3600. * 1000. for k,x in sorted_result],
+    [float(x['content_input']) / x['subsession_length'] * 3600. * 1000. for k,x in sorted_result]
+], np.float32)
+```
+Let's take a look at the parent numbers over time. If the graphs fit tightly together, that's probably a good indicator that they're closely related.
+
+
+```python
+plt.title("Parent Hang Stats")
+plt.xlabel("Build date")
+plt.ylabel("Hangs per kuh")
+
+bhr_index = 0
+input_index = 2
+
+plt.xticks(range(0, len(sorted_result)))
+max_y = max(np.amax(plot_data[bhr_index]), np.amax(plot_data[input_index]))
+plt.yticks(np.arange(0., max_y, max_y / 20.))
+
+plt.grid(True)
+
+plt.plot(range(0, len(sorted_result)), plot_data[bhr_index])
+plt.plot(range(0, len(sorted_result)), plot_data[input_index])
+plt.legend(["bhr", "input"], loc="upper right")
+```
+
+
+
+
+    <matplotlib.legend.Legend at 0x7fd6e1205410>
+
+
+
+
+
+![png](images/output_8_1.png)
+
+
+The parent's data looks _pretty_ close. How about content's?
```

Comment:
What does "close" mean in this context? The average of these two series are pretty close, but they do not appear to change in the same direction. In fact, it looks like they may be a little negatively correlated.

It seems like you want to understand if these series vary together (like the content series do). If that's the case, can you include a correlation coefficient? It would also be useful to show a scatter-plot where your x-axis is `input hangs per kuh` and your y-axis is `bhr hangs per kuh` (i.e. a [q-q plot](https://en.wikipedia.org/wiki/Q%E2%80%93Q_plot). If you only have time for one of these, I'd prioritize the plot.